tardis/energy_input/gamma_ray_packet_source.py

```diff
@@ -827,29 +833,48 @@ def create_packets(
             nus_cmf,
             statuses,
             shells,
-            times,
+            sampled_times.values,
+            positron_energy,
         )
 
-    def calculate_positron_fraction(self, isotopes):
+    def calculate_positron_fraction(self, isotopes, number_of_packets):
         """Calculate the fraction of energy that an isotope
-        releases as positron kinetic energy
+        releases as positron kinetic energy vs the gamma-ray energy
 
         Parameters
         ----------
         isotopes : array
-            Array of isotope names as strings
+            Array of isotope names as strings. Here each isotope is associated with a packet.
+        number_of_packets : int
+            Number of gamma-ray packets
 
         Returns
         -------
         dict
             Fraction of energy released as positron kinetic energy per isotope
         """
-        positron_fraction = {}
-
-        for isotope in isotopes:
-            isotope_energy = self.gamma_ray_lines[isotope][0, :]
-            isotope_intensity = self.gamma_ray_lines[isotope][1, :]
-            positron_fraction[isotope] = self.average_positron_energies[
-                isotope
-            ] / np.sum(isotope_energy * isotope_intensity)
-        return positron_fraction
+        isotope_positron_fraction = np.zeros(number_of_packets)
+
+        # Find the positron fraction from the zeroth shell of the dataframe
+        shell_number_0 = self.isotope_decay_df[
+            self.isotope_decay_df.index.get_level_values("shell_number") == 0
+        ]
+
+        gamma_decay_df = shell_number_0[shell_number_0["radiation"] == "g"]
+
+        positrons_decay_df = shell_number_0[shell_number_0["radiation"] == "bp"]
+        # Find the total energy released from positrons per isotope from the dataframe
+        positron_energy_per_isotope = positrons_decay_df.groupby("isotope")[
+            "energy_per_channel_keV"
+        ].sum()
+        # Find the total energy released from gamma-ray per isotope from the dataframe
```

Comment:
I would add a TODO note here that you may want to change it back to the fraction of total energy instead.

Reply:
Added.